Safe CMA-ES (GECCO 2024)

kento031 opened the pull request:
This is the implementation of Safe CMA-ES [Uchida+, GECCO 2024].
https://arxiv.org/abs/2405.10534

nomuramasahir0 commented:
@c-bata 
This PR introduces a new method called SafeCMA, which will require `gpytorch` and `torch` (for Gaussian processes) when users choose to use it. I believe this PR won’t affect Optuna since SafeCMA is unrelated to Optuna, but could you please double-check to ensure there aren’t any issues?

nomuramasahir0 commented on README.md:
```diff
@@ -463,3 +463,4 @@ Rate Adaptation: Can CMA-ES with Default Population Size Solve Multimodal
 and Noisy Problems?, GECCO, 2023.](https://arxiv.org/abs/2304.03473)
 * [Nomura and Shibata 2024] [M. Nomura, M. Shibata, cmaes : A Simple yet Practical Python Library for CMA-ES, arXiv:2402.01373, 2024.](https://arxiv.org/abs/2402.01373)
 * [Ros and Hansen 2008] [R. Ros, N. Hansen, A Simple Modification in CMA-ES Achieving Linear Time and Space Complexity, PPSN, 2008.](https://hal.inria.fr/inria-00287367/document)
+* [Uchida et al. 2024] [K. Uchida, R. Hamano, M. Nomura, S. Saito, S. Shirakawa, CMA-ES for Safe Optimization, GECCO, 2024.](https://arxiv.org/abs/2405.10534)
```
Could you provide a brief description and usage for SafeCMA (like [LRA-CMA-ES](https://github.com/CyberAgentAILab/cmaes?tab=readme-ov-file#learning-rate-adaptation-cma-es-nomura-et-al-2023))?

nomuramasahir0 replied:
It would be clearer to explicitly state that the current implementation assumes no noise.

nomuramasahir0 commented on cmaes/_safe_cma.py:
```diff
@@ -0,0 +1,702 @@
+
+        cov:
+            A covariance matrix (optional).
+    """
```
Please add the paper link like [this](https://github.com/CyberAgentAILab/cmaes/blob/main/cmaes/_catcma.py#L81).

nomuramasahir0 commented on cmaes/_safe_cma.py:
```diff
@@ -0,0 +1,702 @@
+        assert c1 <= 1 - cmu, "invalid learning rate for the rank-one update"
+        assert cmu <= 1 - c1, "invalid learning rate for the rank-μ update"
+
+        cm = 1  # (eq. 54)
```
I suggest removing comments like eq. ~ (for Hansen's Tutorial paper) in this PR, as they could be confusing when compared with the equations in the SafeCMA paper.

nomuramasahir0 replied:
On the other hand, adding equation numbers for SafeCMA-specific operations would improve clarity.

nomuramasahir0 commented on cmaes/_safe_cma.py:
```diff
@@ -0,0 +1,702 @@
+        likelihood = gpytorch.likelihoods.GaussianLikelihood(
+            noise_constraint=gpytorch.constraints.GreaterThan(0)
+        )
+        likelihood.noise = 0
```
I'm curious about the validity of this setting because, as mentioned in this thread:
> Some small value, but don't make it too small or numerical performance will suffer. I recommend 1e-4.
https://github.com/cornellius-gp/gpytorch/issues/1196#issuecomment-651193968

This implies there's room for improvement in this setting, but I'm not sure for now, so let's see how it goes.

kento031 replied:
This setting was used in some previous studies, including the followings.
https://link.springer.com/chapter/10.1007/978-3-030-58112-1_13
https://dl.acm.org/doi/10.1007/978-3-031-30229-9_51

In addition, I run the safe CMA-ES with `likelihood.noise = 1e-4` on some benchmark functions. As a result, it occurred a few constraint violations that was not observed with `likelihood.noise = 0`. Therefore, I did not change it.

nomuramasahir0 commented on cmaes/_safe_cma.py:
```diff
@@ -0,0 +1,702 @@
+            if out_scalar:
+                grad_norm = grad_norm.mean().to(torch.float64)
+
+            return -grad_norm
```
`df` is defined as `# function that returns the norm of gradient`, but it currently returns a negative value, which seems odd. 
I understand this is intended for `scipy.optimize.minimize`, but it would be better to align it with the description.

kento031 replied:
I changed the explanation of the function as `# function that returns the negative norm of gradient`.

nomuramasahir0 commented on cmaes/_safe_cma.py:
```diff
@@ -0,0 +1,702 @@
+            safe_seeds = (np.random.rand(safe_seeds_num, dim) * 2 - 1) * 5
+            safe_seeds[:, 0] = - np.abs(safe_seeds[:, 0])
+
+            # evaluation of safe seeds (with multiple safety functions)
```
> with multiple safety functions

with a single safety function?

nomuramasahir0 commented on examples/safecma.py:
```diff
@@ -0,0 +1,137 @@
+    safe_seeds = (np.random.rand(safe_seeds_num, dim) * 2 - 1) * 5
+    safe_seeds[:, 0] = -np.abs(safe_seeds[:, 0])
+
+    # evaluation of safe seeds (with multiple safety functions)
```
> with multiple safety functions
with a single safety function?

nomuramasahir0 commented:
@kento031 
Thanks for the PR! I've left some comments. Could you take a look?

c-bata commented on cmaes/__init__.py:
```diff
@@ -6,4 +6,9 @@
 from ._dxnesic import DXNESIC  # NOQA
 from ._catcma import CatCMA  # NOQA
 
+try:
+    from ._safe_cma import SafeCMA  # NOQA
+except ImportError:
+    pass  # Implementation of Safe CMA-ES requires scipy, gpytorch, and torch
+
```
How about simply removing `SafeCMA` and renaming `cmaes/_safe_cma.py` to `cmaes/safe_cma.py`?

```suggestion
```

nomuramasahir0 replied:
Thanks for the great suggestion! It looks good to me.

@kento031 
Could you make this modification if it’s okay?

kento031 commented:
@nomuramasahir0 @c-bata 
Thank you for your kind comments. I will address them.

kento031 commented:
Sorry for being so late in addressing your comments. 
I modified my implementation according to your comments.

I checked the code using `act` (https://github.com/nektos/act), and there remains an error when executing `act -j lint` (and when executing `act pull_request`). This error was caused at `cmaes/_cma.py:480` by running `mypy cmaes`. I leave this error as is for now because I did not change `cmaes/_cma.py`.

@nomuramasahir0 We previously discussed about whether `_safe_cma.py` sould be contained in `cmaes/__init__.py`. However, at that time, I misunderstood how the workflow runs. Finally, I modified my implementation according to @c-bata 's comment.

ha-mano commented:
@kento031 
I have fixed a mypy error that was causing in the workflow tests. The fix is included in [this pull request](https://github.com/CyberAgentAILab/cmaes/pull/195)

kento031 commented:
@ha-mano 
Thank you for fixing the error. I have merged your fix.

kento031 commented on requirements-dev.txt:
```diff
@@ -1,5 +1,7 @@
 # install_requires
 numpy>=1.20.0
+torch
+gpytorch
```
Thank you for your suggestion. I have merged it.

c-bata reviewed (approved):
LGTM. Thank you for your contribution :100:

kento031 commented:
Thank you for all your comments!

In the CI check, the `lint` job failed due to `tools/cmaes_visualizer.py` and `tools/ws_cmaes_visualizer.py`. However, I did not change these files.

Should I modify these codes to resolve the errors?

c-bata commented:
@kento031 No, you don't need to fix that errors in this PR 👍
